tools/x-repo-types/typeAnalyzer/typeAnalyzer.go

```diff
@@ -0,0 +1,575 @@
+// building their TypeNode trees. Stop traversing when a cycle is detected.
+func (t *TypeNode) Build() TypePath {
+	return t.build(TypePath{})
+}
+
+func (t *TypeNode) appendChild(typeName, typeTag string, child TypeNode) {
+	cname := ChildName{typeName, typeTag}
+	t.ChildNames = append(t.ChildNames, cname)
+	if t.children == nil {
+		children := make(ChildTypes)
+		t.children = &children
+	}
+	(*t.children)[cname.String()] = child
+}
+
+// build the TypeNode tree by finding all child types that belong to a kind and recursively
+// building their TypeNode trees.
+// Return a path that will be non-trivial only in a case that a cycle is detected.
+func (t *TypeNode) build(path TypePath) TypePath {
+	if t.IsStruct() {
+		me := t.String()
+		foundCycle := false
+		for _, node := range path {
+			if node.String() == me {
+				foundCycle = true
+				break
+			}
+		}
+		path = append(path, *t)
+		if foundCycle {
+			return path
+		}
+	}
+
+	var cyclePath TypePath
+	switch t.Kind {
+	case reflect.Struct:
+		cyclePath = t.buildStructChildren(path)
+	case reflect.Slice, reflect.Array:
+		cyclePath = t.buildListChild(path)
+	case reflect.Map:
+		cyclePath = t.buildMapChildren(path)
+	case reflect.Ptr:
+		cyclePath = t.buildPtrChild(path)
+	}
+
+	return cyclePath
+}
+
+// buildStructChildren builds the children of a struct type.
+func (t *TypeNode) buildStructChildren(path TypePath) TypePath {
+	var cyclePath TypePath
+	for i := 0; i < t.Type.NumField(); i++ {
+		typeField := t.Type.Field(i)
+		typeName := typeField.Name
+
+		// probably we need to skip typeField.Tag == `codec:"-"` as well
+		if typeName == "" || (!unicode.IsUpper(rune(typeName[0])) && typeName != "_struct") {
+			continue
+		}
+
+		if typeField.Anonymous {
+			// embedded struct case
+			actualKind := typeField.Type.Kind()
+			if actualKind != reflect.Struct {
+				panic(fmt.Sprintf("expected [%s] but got unexpected embedded type: %s", reflect.Struct, typeField.Type))
+			}
+
+			embedded := TypeNode{t.Depth, typeField.Type, reflect.Struct, nil, nil}
+			embeddedCyclePath := embedded.build(path)
+			if len(embeddedCyclePath) > 0 {
+				cyclePath = embeddedCyclePath
+			}
+			for _, edge := range embedded.ChildNames {
+				child := (*embedded.children)[edge.String()]
+				t.appendChild(edge.Name, edge.Tag, child)
+			}
+			continue
+		}
+
+		typeTag := string(typeField.Tag)
+		child := TypeNode{t.Depth + 1, typeField.Type, typeField.Type.Kind(), nil, nil}
+		childCyclePath := child.build(path)
+		if len(childCyclePath) > 0 {
+			cyclePath = childCyclePath
+		}
+		t.appendChild(typeName, typeTag, child)
+	}
+	return cyclePath
+}
+
+func (t *TypeNode) buildListChild(path TypePath) TypePath {
+	tt := t.Type.Elem()
+	child := TypeNode{t.Depth + 1, tt, tt.Kind(), nil, nil}
+	path = child.build(path)
+	t.appendChild("<list elt>", "", child)
+	return path
+}
+
+func (t *TypeNode) buildMapChildren(path TypePath) TypePath {
```

Comment:
Nice trick, I especially like that you were able to specify that it is a key/val with a special name :+1: 
A comment saying that maps are handled by adding the key and the value as children would be helpful.

Reply:
55e1d9a7f6673ecc954f632e94795245e1fe1902